src/io/geotiff.rs

```diff
@@ -18,6 +20,55 @@ impl<R: Read + Seek> CogReader<R> {
 
         Ok(Self { decoder })
     }
+
+    /// Affine transformation for 2D matrix extracted from TIFF tag metadata, used to transform
+    /// image pixel (row, col) coordinates to and from geographic/projected (x, y) coordinates.
+    ///
+    /// ```text
+    /// | x' |   | a b c | | x |
+    /// | y' | = | d e f | | y |
+    /// | 1  |   | 0 0 1 | | 1 |
+    /// ```
+    ///
+    /// where (`x'` and `y'`) are world coordinates, and (`x`, `y) are the pixel's image
+    /// coordinates. Letters a to f represent:
+    ///
+    /// - `a` - width of a pixel (x-resolution)
+    /// - `b` - row rotation (typically zero)
+    /// - `c` - x-coordinate of the *center* of the upper-left pixel (x-origin)
+    /// - `d` - column rotation (typically zero)
+    /// - `e` - height of a pixel (y-resolution, typically negative)
+    /// - `f` - y-coordinate of the *center* of the upper-left pixel (y-origin)
+    ///
+    /// References:
+    /// - <https://docs.ogc.org/is/19-008r4/19-008r4.html#_coordinate_transformations>
+    fn transform(&mut self) -> TiffResult<AffineTransform<f64>> {
+        // Get x and y axis rotation (not yet implemented)
+        let (x_rotation, y_rotation): (f64, f64) =
+            match self.decoder.get_tag_f64_vec(Tag::ModelTransformationTag) {
+                Ok(_model_transformation) => unimplemented!("Non-zero rotation is not handled yet"),
+                Err(_) => (0.0, 0.0),
+            };
+
+        // Get pixel size in x and y direction
+        let pixel_scale: Vec<f64> = self.decoder.get_tag_f64_vec(Tag::ModelPixelScaleTag)?;
+        let [x_scale, y_scale, _z_scale] = pixel_scale[0..3] else {
+            return Err(TiffError::FormatError(TiffFormatError::InvalidTag));
+        };
+
+        // Get x and y coordinates of upper left pixel
+        let tie_points: Vec<f64> = self.decoder.get_tag_f64_vec(Tag::ModelTiepointTag)?;
+        let [_i, _j, _k, x_origin, y_origin, _z_origin] = tie_points[0..6] else {
+            return Err(TiffError::FormatError(TiffFormatError::InvalidTag));
+        };
```

Comment:
Tried a few different ways to convert the Vec to an array (e.g. following https://stackoverflow.com/questions/29570607/is-there-a-good-way-to-convert-a-vect-to-an-array), but couldn't quite get it to work, and the code was very messy. The code here works, but might silently handle cases where `pixel_scale.len() > 3` or `tie_points.len() > 6` incorrectly, rather than raising an error, so should come back and refactor this sometime.